include/swift/AST/Types.h

```diff
@@ -1025,6 +1025,76 @@ class alignas(1 << TypeAlignInBits) TypeBase {
   void *operator new(size_t Bytes, void *Mem) throw() { return Mem; }
 };
 
+/// AnyGenericType - This abstract class helps types ensure that fields
+/// exist at the same offset in memory to improve code generation of the
+/// compiler itself.
+class AnyGenericType : public TypeBase {
+  friend class NominalOrBoundGenericNominalType;
+
+  /// TheDecl - This is the TypeDecl which declares the given type. It
+  /// specifies the name and other useful information about this type.
+  union {
+    GenericTypeDecl *GenDecl;
+    NominalTypeDecl *NomDecl;
+  };
```

Comment:
That's a clever bit of aliasing here... but it'll work fine.

Reply:
Thanks. Ya, the union is gross. I tried using templates, but that fought with the CanType macros. I wasn't sure what was more important. It should be polished at some point.